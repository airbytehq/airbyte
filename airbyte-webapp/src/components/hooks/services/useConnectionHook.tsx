import { useFetcher, useResource } from "rest-hooks";

import config from "../../../config";
import { AnalyticsService } from "../../../core/analytics/AnalyticsService";
import ConnectionResource, {
  Connection
} from "../../../core/resources/Connection";
import WorkspaceResource from "../../../core/resources/Workspace";
import { SyncSchema } from "../../../core/resources/Schema";
import { Source } from "../../../core/resources/Source";
import { Destination } from "../../../core/resources/Destination";
import FrequencyConfig from "../../../data/FrequencyConfig.json";
import { SourceImplementation } from "../../../core/resources/SourceImplementation";

type ValuesProps = {
  frequency: string;
  syncSchema: SyncSchema;
  source?: Source;
};

type CreateConnectionProps = {
  values: ValuesProps;
  sourceImplementation?: SourceImplementation;
  destinationImplementationId: string;
  sourceConnector?: Source;
  destinationConnector?: Destination;
};

const useConnection = () => {
  const createConnectionResource = useFetcher(ConnectionResource.createShape());
  const updateWorkspace = useFetcher(WorkspaceResource.updateShape());
  const workspace = useResource(WorkspaceResource.detailShape(), {
    workspaceId: config.ui.workspaceId
  });
  const updateConnectionResource = useFetcher(ConnectionResource.updateShape());
  const updateStateConnectionResource = useFetcher(
    ConnectionResource.updateStateShape()
  );
  const createConnection = async ({
    values,
    sourceImplementation,
    destinationImplementationId,
    sourceConnector,
    destinationConnector
  }: CreateConnectionProps) => {
    const frequencyData = FrequencyConfig.find(
      item => item.value === values.frequency
    );

    try {
      const result = await createConnectionResource(
        {
          sourceId: values.source?.sourceId || "",
          sourceName: values.source?.name || "",
          name: sourceImplementation?.name || ""
        },
        {
          sourceImplementationId: sourceImplementation?.sourceImplementationId,
          destinationImplementationId,
          syncMode: "full_refresh",
          schedule: frequencyData?.config,
          status: "active",
          syncSchema: values.syncSchema
        },
        [
          [
            ConnectionResource.listShape(),
            { workspaceId: config.ui.workspaceId },
            (
              newConnectionId: string,
              connectionsIds: { connections: string[] }
            ) => ({
              connections: [
                ...(connectionsIds?.connections || []),
                newConnectionId
              ]
            })
          ]
        ]
      );
      AnalyticsService.track("New Connection - Action", {
        user_id: config.ui.workspaceId,
        action: "Set up connection",
        frequency: frequencyData?.text,
        connector_source: sourceConnector?.name,
        connector_source_id: sourceConnector?.sourceId,
        connector_destination: destinationConnector?.name,
        connector_destination_id: destinationConnector?.destinationId
      });

      if (!workspace.onboardingComplete) {
        await updateWorkspace(
          {},
          {
            workspaceId: workspace.workspaceId,
            initialSetupComplete: workspace.initialSetupComplete,
            onboardingComplete: true,
            anonymousDataCollection: workspace.anonymousDataCollection,
            news: workspace.news,
            securityUpdates: workspace.securityUpdates
          }
        );
      }

      return result;
    } catch (e) {
      throw e;
    }
  };

  const updateConnection = async ({
    connectionId,
    syncSchema,
    status,
    schedule
  }: {
    connectionId: string;
    syncSchema?: SyncSchema;
    status: string;
    schedule: {
      units: number;
      timeUnit: string;
    } | null;
  }) => {
    await updateConnectionResource(
      {},
      {
        connectionId,
        syncSchema,
        status,
        schedule
      }
    );
  };

  const updateStateConnection = async ({
    sourceData,
    sourceName,
    connectionConfiguration,
    schedule
  }: {
    sourceData: Connection;
    sourceName: string;
    connectionConfiguration: any;
    schedule: {
      units: number;
      timeUnit: string;
    } | null;
  }) => {
    await updateStateConnectionResource(
      {},
      {
        ...sourceData,
        schedule,
        source: {
          ...sourceData.source,
          name: sourceName,
          connectionConfiguration: connectionConfiguration
        }
      }
    );
  };

  return {
    createConnection,
    updateConnection,
    updateStateConnection
  };
};

export default useConnection;
