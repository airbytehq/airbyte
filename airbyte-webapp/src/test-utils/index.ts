export * from "./testutils";
export * from "./TestSuspenseBoundary";
