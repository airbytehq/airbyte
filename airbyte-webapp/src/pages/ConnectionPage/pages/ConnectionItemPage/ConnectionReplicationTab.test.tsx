/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { render as tlr, act } from "@testing-library/react";
import React, { Suspense } from "react";
import mockConnection from "test-utils/mock-data/mockConnection.json";
import mockDest from "test-utils/mock-data/mockDestinationDefinition.json";
import { TestWrapper } from "test-utils/testutils";

import { WebBackendConnectionUpdate } from "core/request/AirbyteClient";
import { ConnectionEditServiceProvider } from "hooks/services/ConnectionEdit/ConnectionEditService";
import * as connectionHook from "hooks/services/useConnectionHook";

import { ConnectionReplicationTab } from "./ConnectionReplicationTab";

jest.mock("services/connector/DestinationDefinitionSpecificationService", () => ({
  useGetDestinationDefinitionSpecification: () => mockDest,
}));

describe("ConnectionReplicationTab", () => {
  const Wrapper: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => (
    <Suspense fallback={<div>I should not show up in a snapshot</div>}>
      <TestWrapper>
        <ConnectionEditServiceProvider connectionId={mockConnection.connectionId}>
          {children}
        </ConnectionEditServiceProvider>
      </TestWrapper>
    </Suspense>
  );
  const render = async () => {
    let renderResult: ReturnType<typeof tlr>;
    await act(async () => {
      renderResult = tlr(
        <Wrapper>
          <ConnectionReplicationTab />
        </Wrapper>
      );
    });
    return renderResult!;
  };

  const setupSpies = (getConnection?: () => Promise<void>) => {
    const getConnectionImpl: any = {
      getConnection: getConnection ?? (() => new Promise(() => null) as any),
    };
    jest.spyOn(connectionHook, "useGetConnection").mockImplementation(() => mockConnection as any);
    jest.spyOn(connectionHook, "useWebConnectionService").mockImplementation(() => getConnectionImpl);
    jest.spyOn(connectionHook, "useUpdateConnection").mockImplementation(
      () =>
        ({
          mutateAsync: async (connection: WebBackendConnectionUpdate) => connection,
          isLoading: false,
        } as any)
    );
  };
  it("should render", async () => {
    setupSpies();

    const renderResult = await render();
    expect(renderResult).toMatchSnapshot();
  });

  it("should show an error if there is a schemaError", async () => {
    setupSpies(() => Promise.reject("Test Error"));

    const renderResult = await render();

    await act(async () => {
      renderResult.queryByText("Refresh source schema")?.click();
    });
    expect(renderResult).toMatchSnapshot();
  });

  it("should show loading if the schema is refreshing", async () => {
    setupSpies();

    const renderResult = await render();
    await act(async () => {
      renderResult.queryByText("Refresh source schema")?.click();
    });

    await act(async () => {
      expect(renderResult.findByText("We are fetching the schema of your data source.", { exact: false })).toBeTruthy();
    });
  });
});
