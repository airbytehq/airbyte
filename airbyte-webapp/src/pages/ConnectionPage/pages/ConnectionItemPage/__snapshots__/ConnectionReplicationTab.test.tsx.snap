// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionReplicationTab should render 1`] = `
<body>
  <div>
    <div
      class="<removed-for-snapshot-test>"
    >
      <form
        action="#"
      >
        <div
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
          >
            <div
              class="<removed-for-snapshot-test>"
            >
              <h2
                class="<removed-for-snapshot-test>"
              >
                Transfer
              </h2>
              <div
                class="<removed-for-snapshot-test>"
              >
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <label
                      class="<removed-for-snapshot-test>"
                    >
                      Replication frequency*
                      <span>
                        <br />
                        <span
                          class="<removed-for-snapshot-test>"
                        >
                          Set how often data should sync to the destination
                        </span>
                      </span>
                    </label>
                  </div>
                </div>
                <div
                  class="<removed-for-snapshot-test>"
                  style="pointer-events: auto;"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                    data-testid="scheduleData"
                    role="combobox"
                  >
                    <span
                      class="<removed-for-snapshot-test>"
                      id="react-select-2-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="<removed-for-snapshot-test>"
                    />
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <div
                            class="<removed-for-snapshot-test>"
                          >
                            <div
                              class="<removed-for-snapshot-test>"
                            >
                              Manual
                            </div>
                          </div>
                        </div>
                        <input
                          aria-autocomplete="list"
                          aria-expanded="false"
                          aria-haspopup="true"
                          aria-readonly="true"
                          class="<removed-for-snapshot-test>"
                          id="react-select-2-input"
                          inputmode="none"
                          role="combobox"
                          tabindex="0"
                          value=""
                        />
                      </div>
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          aria-hidden="true"
                          class="<removed-for-snapshot-test>"
                        >
                          <svg
                            aria-hidden="true"
                            class="<removed-for-snapshot-test>"
                            data-icon="sort-down"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 320 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M311.9 335.1l-132.4 136.8C174.1 477.3 167.1 480 160 480c-7.055 0-14.12-2.702-19.47-8.109l-132.4-136.8C-9.229 317.8 3.055 288 27.66 288h264.7C316.9 288 329.2 317.8 311.9 335.1z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <input
                      name="scheduleData"
                      type="hidden"
                      value="manual"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="<removed-for-snapshot-test>"
          >
            <div
              class="<removed-for-snapshot-test>"
            >
              <h2
                class="<removed-for-snapshot-test>"
              >
                Streams
              </h2>
              <span
                class=""
              >
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <label
                        class="<removed-for-snapshot-test>"
                      >
                        Destination Namespace*
                        <span>
                          <br />
                          <span
                            class="<removed-for-snapshot-test>"
                          >
                            Define the location where the data will be stored in the destination
                          </span>
                        </span>
                      </label>
                    </div>
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <div
                      class="<removed-for-snapshot-test>"
                      data-testid="namespaceDefinition"
                      role="combobox"
                    >
                      <span
                        class="<removed-for-snapshot-test>"
                        id="react-select-3-live-region"
                      />
                      <span
                        aria-atomic="false"
                        aria-live="polite"
                        aria-relevant="additions text"
                        class="<removed-for-snapshot-test>"
                      />
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <div
                            class="<removed-for-snapshot-test>"
                          >
                            <div
                              class="<removed-for-snapshot-test>"
                            >
                              <div
                                class="<removed-for-snapshot-test>"
                              >
                                Mirror source structure
                              </div>
                            </div>
                          </div>
                          <input
                            aria-autocomplete="list"
                            aria-expanded="false"
                            aria-haspopup="true"
                            aria-readonly="true"
                            class="<removed-for-snapshot-test>"
                            id="react-select-3-input"
                            inputmode="none"
                            role="combobox"
                            tabindex="0"
                            value=""
                          />
                        </div>
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <div
                            aria-hidden="true"
                            class="<removed-for-snapshot-test>"
                          >
                            <svg
                              aria-hidden="true"
                              class="<removed-for-snapshot-test>"
                              data-icon="sort-down"
                              data-prefix="fas"
                              focusable="false"
                              role="img"
                              viewBox="0 0 320 512"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M311.9 335.1l-132.4 136.8C174.1 477.3 167.1 480 160 480c-7.055 0-14.12-2.702-19.47-8.109l-132.4-136.8C-9.229 317.8 3.055 288 27.66 288h264.7C316.9 288 329.2 317.8 311.9 335.1z"
                                fill="currentColor"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                      <input
                        name="namespaceDefinition"
                        type="hidden"
                        value="source"
                      />
                    </div>
                  </div>
                </div>
              </span>
              <div
                class="<removed-for-snapshot-test>"
              >
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <label
                      class="<removed-for-snapshot-test>"
                    >
                      Destination Stream Prefix (Optional)
                      <span>
                        <br />
                        <span
                          class="<removed-for-snapshot-test>"
                        >
                          Add a prefix to stream names (ex. “airbyte_” causes “projects” =&gt; “airbyte_projects”)
                        </span>
                      </span>
                    </label>
                  </div>
                </div>
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                    data-testid="input-container"
                  >
                    <input
                      class="<removed-for-snapshot-test>"
                      data-testid="prefixInput"
                      name="prefix"
                      placeholder="prefix"
                      style="pointer-events: auto;"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="<removed-for-snapshot-test>"
          >
            <div
              class="<removed-for-snapshot-test>"
            >
              <div
                class="<removed-for-snapshot-test>"
              >
                <h2
                  class="<removed-for-snapshot-test>"
                >
                  Activate the streams you want to sync
                </h2>
                <button
                  class="<removed-for-snapshot-test>"
                  type="button"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <svg
                      aria-hidden="true"
                      class="<removed-for-snapshot-test>"
                      data-icon="rotate"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M449.9 39.96l-48.5 48.53C362.5 53.19 311.4 32 256 32C161.5 32 78.59 92.34 49.58 182.2c-5.438 16.81 3.797 34.88 20.61 40.28c16.97 5.5 34.86-3.812 40.3-20.59C130.9 138.5 189.4 96 256 96c37.96 0 73 14.18 100.2 37.8L311.1 178C295.1 194.8 306.8 223.4 330.4 224h146.9C487.7 223.7 496 215.3 496 204.9V59.04C496 34.99 466.9 22.95 449.9 39.96zM441.8 289.6c-16.94-5.438-34.88 3.812-40.3 20.59C381.1 373.5 322.6 416 256 416c-37.96 0-73-14.18-100.2-37.8L200 334C216.9 317.2 205.2 288.6 181.6 288H34.66C24.32 288.3 16 296.7 16 307.1v145.9c0 24.04 29.07 36.08 46.07 19.07l48.5-48.53C149.5 458.8 200.6 480 255.1 480c94.45 0 177.4-60.34 206.4-150.2C467.9 313 458.6 294.1 441.8 289.6z"
                        fill="currentColor"
                      />
                    </svg>
                    Refresh source schema
                  </div>
                </button>
              </div>
              <div
                class="<removed-for-snapshot-test>"
              >
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                    data-testid="input-container"
                  >
                    <input
                      class="<removed-for-snapshot-test>"
                      data-testid="input"
                      placeholder="Search stream name"
                    />
                  </div>
                </div>
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <label
                      class="<removed-for-snapshot-test>"
                    >
                      <input
                        class="<removed-for-snapshot-test>"
                        type="checkbox"
                      />
                    </label>
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  />
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Sync
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Source
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <svg
                            fill="none"
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                          >
                            <path
                              d="M7.00016 13.6663C3.31816 13.6663 0.333496 10.6817 0.333496 6.99967C0.333496 3.31767 3.31816 0.333008 7.00016 0.333008C10.6822 0.333008 13.6668 3.31767 13.6668 6.99967C13.6668 10.6817 10.6822 13.6663 7.00016 13.6663ZM7.00016 12.333C8.41465 12.333 9.77121 11.7711 10.7714 10.7709C11.7716 9.77072 12.3335 8.41416 12.3335 6.99967C12.3335 5.58519 11.7716 4.22863 10.7714 3.22844C9.77121 2.22824 8.41465 1.66634 7.00016 1.66634C5.58567 1.66634 4.22912 2.22824 3.22893 3.22844C2.22873 4.22863 1.66683 5.58519 1.66683 6.99967C1.66683 8.41416 2.22873 9.77072 3.22893 10.7709C4.22912 11.7711 5.58567 12.333 7.00016 12.333ZM6.3335 3.66634H7.66683V4.99967H6.3335V3.66634ZM6.3335 6.33301H7.66683V10.333H6.3335V6.33301Z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  />
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Sync mode
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <svg
                            fill="none"
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                          >
                            <path
                              d="M7.00016 13.6663C3.31816 13.6663 0.333496 10.6817 0.333496 6.99967C0.333496 3.31767 3.31816 0.333008 7.00016 0.333008C10.6822 0.333008 13.6668 3.31767 13.6668 6.99967C13.6668 10.6817 10.6822 13.6663 7.00016 13.6663ZM7.00016 12.333C8.41465 12.333 9.77121 11.7711 10.7714 10.7709C11.7716 9.77072 12.3335 8.41416 12.3335 6.99967C12.3335 5.58519 11.7716 4.22863 10.7714 3.22844C9.77121 2.22824 8.41465 1.66634 7.00016 1.66634C5.58567 1.66634 4.22912 2.22824 3.22893 3.22844C2.22873 4.22863 1.66683 5.58519 1.66683 6.99967C1.66683 8.41416 2.22873 9.77072 3.22893 10.7709C4.22912 11.7711 5.58567 12.333 7.00016 12.333ZM6.3335 3.66634H7.66683V4.99967H6.3335V3.66634ZM6.3335 6.33301H7.66683V10.333H6.3335V6.33301Z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Cursor field
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <svg
                            fill="none"
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                          >
                            <path
                              d="M7.00016 13.6663C3.31816 13.6663 0.333496 10.6817 0.333496 6.99967C0.333496 3.31767 3.31816 0.333008 7.00016 0.333008C10.6822 0.333008 13.6668 3.31767 13.6668 6.99967C13.6668 10.6817 10.6822 13.6663 7.00016 13.6663ZM7.00016 12.333C8.41465 12.333 9.77121 11.7711 10.7714 10.7709C11.7716 9.77072 12.3335 8.41416 12.3335 6.99967C12.3335 5.58519 11.7716 4.22863 10.7714 3.22844C9.77121 2.22824 8.41465 1.66634 7.00016 1.66634C5.58567 1.66634 4.22912 2.22824 3.22893 3.22844C2.22873 4.22863 1.66683 5.58519 1.66683 6.99967C1.66683 8.41416 2.22873 9.77072 3.22893 10.7709C4.22912 11.7711 5.58567 12.333 7.00016 12.333ZM6.3335 3.66634H7.66683V4.99967H6.3335V3.66634ZM6.3335 6.33301H7.66683V10.333H6.3335V6.33301Z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Primary key
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <svg
                            fill="none"
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                          >
                            <path
                              d="M7.00016 13.6663C3.31816 13.6663 0.333496 10.6817 0.333496 6.99967C0.333496 3.31767 3.31816 0.333008 7.00016 0.333008C10.6822 0.333008 13.6668 3.31767 13.6668 6.99967C13.6668 10.6817 10.6822 13.6663 7.00016 13.6663ZM7.00016 12.333C8.41465 12.333 9.77121 11.7711 10.7714 10.7709C11.7716 9.77072 12.3335 8.41416 12.3335 6.99967C12.3335 5.58519 11.7716 4.22863 10.7714 3.22844C9.77121 2.22824 8.41465 1.66634 7.00016 1.66634C5.58567 1.66634 4.22912 2.22824 3.22893 3.22844C2.22873 4.22863 1.66683 5.58519 1.66683 6.99967C1.66683 8.41416 2.22873 9.77072 3.22893 10.7709C4.22912 11.7711 5.58567 12.333 7.00016 12.333ZM6.3335 3.66634H7.66683V4.99967H6.3335V3.66634ZM6.3335 6.33301H7.66683V10.333H6.3335V6.33301Z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Destination
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <svg
                            fill="none"
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                          >
                            <path
                              d="M7.00016 13.6663C3.31816 13.6663 0.333496 10.6817 0.333496 6.99967C0.333496 3.31767 3.31816 0.333008 7.00016 0.333008C10.6822 0.333008 13.6668 3.31767 13.6668 6.99967C13.6668 10.6817 10.6822 13.6663 7.00016 13.6663ZM7.00016 12.333C8.41465 12.333 9.77121 11.7711 10.7714 10.7709C11.7716 9.77072 12.3335 8.41416 12.3335 6.99967C12.3335 5.58519 11.7716 4.22863 10.7714 3.22844C9.77121 2.22824 8.41465 1.66634 7.00016 1.66634C5.58567 1.66634 4.22912 2.22824 3.22893 3.22844C2.22873 4.22863 1.66683 5.58519 1.66683 6.99967C1.66683 8.41416 2.22873 9.77072 3.22893 10.7709C4.22912 11.7711 5.58567 12.333 7.00016 12.333ZM6.3335 3.66634H7.66683V4.99967H6.3335V3.66634ZM6.3335 6.33301H7.66683V10.333H6.3335V6.33301Z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  />
                </div>
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Namespace
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Stream name
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Source | Destination
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  />
                  <div
                    class="<removed-for-snapshot-test>"
                  />
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Namespace
                  </div>
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    Stream name
                  </div>
                </div>
                <div
                  class="<removed-for-snapshot-test>"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <div
                      class="<removed-for-snapshot-test>"
                    >
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <label
                          class="<removed-for-snapshot-test>"
                        >
                          <input
                            class="<removed-for-snapshot-test>"
                            type="checkbox"
                          />
                        </label>
                      </div>
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <span
                          class="<removed-for-snapshot-test>"
                        >
                          <svg
                            aria-hidden="true"
                            class="<removed-for-snapshot-test>"
                            data-icon="chevron-right"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 320 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M96 480c-8.188 0-16.38-3.125-22.62-9.375c-12.5-12.5-12.5-32.75 0-45.25L242.8 256L73.38 86.63c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0l192 192c12.5 12.5 12.5 32.75 0 45.25l-192 192C112.4 476.9 104.2 480 96 480z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </div>
                      <div
                        class="<removed-for-snapshot-test>"
                      >
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <label
                            class="<removed-for-snapshot-test>"
                          >
                            <input
                              checked=""
                              class="<removed-for-snapshot-test>"
                              type="checkbox"
                              value=""
                            />
                            <span
                              class="<removed-for-snapshot-test>"
                            />
                          </label>
                        </div>
                        <div
                          class="<removed-for-snapshot-test>"
                          title=""
                        >
                          <span
                            class="<removed-for-snapshot-test>"
                          >
                            No namespace
                          </span>
                        </div>
                        <div
                          class="<removed-for-snapshot-test>"
                          title="pokemon"
                        >
                          pokemon
                        </div>
                        <div
                          class="<removed-for-snapshot-test>"
                        >
                          <div
                            class="<removed-for-snapshot-test>"
                            data-testid="syncSettingsDropdown"
                            role="combobox"
                          >
                            <span
                              class="<removed-for-snapshot-test>"
                              id="react-select-4-live-region"
                            />
                            <span
                              aria-atomic="false"
                              aria-live="polite"
                              aria-relevant="additions text"
                              class="<removed-for-snapshot-test>"
                            />
                            <div
                              class="<removed-for-snapshot-test>"
                            >
                              <div
                                class="<removed-for-snapshot-test>"
                              >
                                <div
                                  class="<removed-for-snapshot-test>"
                                >
                                  <div
                                    class="<removed-for-snapshot-test>"
                                  >
                                    <div>
                                      Full refresh
                                    </div>
                                    <div
                                      class="<removed-for-snapshot-test>"
                                    >
                                      |
                                    </div>
                                    <div>
                                      Append
                                    </div>
                                  </div>
                                </div>
                                <input
                                  aria-autocomplete="list"
                                  aria-expanded="false"
                                  aria-haspopup="true"
                                  aria-readonly="true"
                                  class="<removed-for-snapshot-test>"
                                  id="react-select-4-input"
                                  inputmode="none"
                                  role="combobox"
                                  tabindex="0"
                                  value=""
                                />
                              </div>
                              <div
                                class="<removed-for-snapshot-test>"
                              >
                                <div
                                  aria-hidden="true"
                                  class="<removed-for-snapshot-test>"
                                >
                                  <svg
                                    aria-hidden="true"
                                    class="<removed-for-snapshot-test>"
                                    data-icon="sort-down"
                                    data-prefix="fas"
                                    focusable="false"
                                    role="img"
                                    viewBox="0 0 320 512"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M311.9 335.1l-132.4 136.8C174.1 477.3 167.1 480 160 480c-7.055 0-14.12-2.702-19.47-8.109l-132.4-136.8C-9.229 317.8 3.055 288 27.66 288h264.7C316.9 288 329.2 317.8 311.9 335.1z"
                                      fill="currentColor"
                                    />
                                  </svg>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="<removed-for-snapshot-test>"
                        />
                        <div
                          class="<removed-for-snapshot-test>"
                        />
                        <div
                          class="<removed-for-snapshot-test>"
                          title="'<source schema>"
                        >
                          '&lt;source schema&gt;
                        </div>
                        <div
                          class="<removed-for-snapshot-test>"
                          title="pokemon"
                        >
                          pokemon
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
          >
            <button
              class="<removed-for-snapshot-test>"
              disabled=""
              type="button"
            >
              <div
                class="<removed-for-snapshot-test>"
              >
                Cancel
              </div>
            </button>
            <button
              class="<removed-for-snapshot-test>"
              disabled=""
              type="submit"
            >
              <div
                class="<removed-for-snapshot-test>"
              >
                Save changes
              </div>
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</body>
`;

exports[`ConnectionReplicationTab should show an error if there is a schemaError 1`] = `
<body>
  <div>
    <div
      class="<removed-for-snapshot-test>"
    >
      <div
        class="<removed-for-snapshot-test>"
      >
        <div
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
          >
            <svg
              aria-hidden="true"
              class="<removed-for-snapshot-test>"
              data-icon="xmark"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M310.6 361.4c12.5 12.5 12.5 32.75 0 45.25C304.4 412.9 296.2 416 288 416s-16.38-3.125-22.62-9.375L160 301.3L54.63 406.6C48.38 412.9 40.19 416 32 416S15.63 412.9 9.375 406.6c-12.5-12.5-12.5-32.75 0-45.25l105.4-105.4L9.375 150.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L160 210.8l105.4-105.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-105.4 105.4L310.6 361.4z"
                fill="currentColor"
              />
            </svg>
          </div>
          <p
            class="<removed-for-snapshot-test>"
          >
            Failed to fetch schema. Please try again
          </p>
          <button
            class="<removed-for-snapshot-test>"
          >
            <div
              class="<removed-for-snapshot-test>"
            >
              Try again
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
